src/y0/mutate/canonicalize.py

```diff
@@ -0,0 +1,108 @@
+
+
+class Canonicalizer:
+    """A data structure to support application of the canonicalize algorithm."""
+
+    def __init__(self, ordering: Sequence[Variable]) -> None:
+        """Initialize the canonicalizer.
+
+        :param ordering: A topological ordering over the variables appearing in the expression.
+        """
+        self.ordering = ordering
+        self.ordering_level = {
+            variable: level
+            for level, variable in enumerate(self.ordering)
+        }
+
+    def _canonicalize_probability(self, expression: Probability) -> Probability:
+        return Probability(Distribution(
+            children=expression.distribution.children,
+            parents=tuple(sorted(expression.distribution.parents, key=self.ordering_level.__getitem__)),
+        ))
+
+    def canonicalize(self, expression: Expression) -> Expression:
+        """Canonicalize an expression.
+
+        :param expression: An uncanonicalized expression
+        :return: A canonicalized expression
+        :raises TypeError: if an object with an invalid type is passed
+        """
+        if isinstance(expression, Probability):  # atomic
+            return self._canonicalize_probability(expression)
+        elif isinstance(expression, Sum):
+            if isinstance(expression.expression, Probability):  # also atomic
+                return expression
+
+            return Sum(
+                expression=self.canonicalize(expression.expression),
+                ranges=expression.ranges,
+            )
+        elif isinstance(expression, Product):
+            probabilities = []
+            other = []
+            for subexpr in expression.expressions:
+                subexpr = self.canonicalize(subexpr)
+                if isinstance(subexpr, Probability):
+                    probabilities.append(subexpr)
+                else:
+                    other.append(subexpr)
+            probabilities = sorted(probabilities, key=_sort_probability_key)
+            other = sorted(other, key=self._nonatomic_key)
+            return Product((*probabilities, *other))
+        elif isinstance(expression, Fraction):
+            return Fraction(
+                numerator=self.canonicalize(expression.numerator),
+                denominator=self.canonicalize(expression.denominator),
+            )
+        else:
+            raise TypeError
+
+    def _nonatomic_key(self, expression: Expression):
+        if isinstance(expression, Probability):
+            return 0, expression.distribution.children[0].name
+        elif isinstance(expression, Product):
+            inner_keys = [
+                self._nonatomic_key(sexpr)
+                for sexpr in expression.expressions
+            ]
+            return 1, *inner_keys
+        elif isinstance(expression, Sum):
+            return 1, self._nonatomic_key(expression.expression)
+
+        raise NotImplementedError('nonatomic sort not implemented')
```

Comment:
I'm pretty sure it's not possible to run into this scenario since a fraction can never appear in a product

Reply:
This was not correct, lol